File: Data-Structures-And-Algorithms/Source-Code/rsa.cpp
#include <iostream>
#include <cmath>

using namespace std;

// Hàm kiểm tra số nguyên tố
bool isPrime(int n) {
    if (n <= 1) return false;
    if (n <= 3) return true;
    if (n % 2 == 0 || n % 3 == 0) return false;
    for (int i = 5; i * i <= n; i = i + 6)
        if (n % i == 0 || n % (i + 2) == 0)
            return false;
    return true;
}

// Hàm tìm số nguyên tố ngẫu nhiên
int findRandomPrime() {
    int num;
    do {
        num = rand() % 100 + 1;
    } while (!isPrime(num));
    return num;
}

// Hàm tìm ước số chung lớn nhất
int gcd(int a, int b) {
    if (b == 0)
        return a;
    return gcd(b, a % b);
}

// Hàm tính toán khóa công khai và khóa riêng tư
void generateKeys(int& publicKey, int& privateKey, int& n) {
    int p = findRandomPrime();
    int q = findRandomPrime();

    n = p * q;
    int phi = (p - 1) * (q - 1);

    publicKey = rand() % (phi - 2) + 2;
    while (gcd(publicKey, phi) != 1) {
        publicKey = rand() % (phi - 2) + 2;
    }

    // Tìm khóa riêng tư dựa trên khóa công khai và mô đun phi
    int k = 2;
    while (((k * phi) + 1) % publicKey != 0) {
        k++;
    }
    privateKey = ((k * phi) + 1) / publicKey;
}

// Hàm mã hóa
int encrypt(int plaintext, int publicKey, int n) {
    return int(pow(plaintext, publicKey)) % n;
}

// Hàm giải mã
int decrypt(int ciphertext, int privateKey, int n) {
    return int(pow(ciphertext, privateKey)) % n;
}

int main() {
    srand(time(NULL));

    int publicKey, privateKey, n;
    generateKeys(publicKey, privateKey, n);

    int plaintext;
    cout << "Nhập vào plaintext: ";
    cin >> plaintext;

    int ciphertext = encrypt(plaintext, publicKey, n);
    cout << "Ciphertext: " << ciphertext << endl;

    int decryptedText = decrypt(ciphertext, privateKey, n);
    cout << "Plaintext sau khi giải mã: " << decryptedText << endl;

    return 0;
}

File: Data-Structures-And-Algorithms/Source-Code/rsa.py
import random
from sympy import mod_inverse, gcd

def generate_keypair(p, q):
    n = p * q
    phi = (p - 1) * (q - 1)
    
    e = random.randint(2, phi - 1)
    while gcd(e, phi) != 1:
        e = random.randint(2, phi - 1)
    
    d = mod_inverse(e, phi)
    
    return ((e, n), (d, n))

def encrypt(message, public_key):
    e, n = public_key
    cipher_text = [pow(ord(char), e, n) for char in message]
    return cipher_text

def decrypt(cipher_text, private_key):
    d, n = private_key
    plain_text = [chr(pow(char, d, n)) for char in cipher_text]
    return ''.join(plain_text)

# Ví dụ về sử dụng RSA
p = 73
q = 89
public_key, private_key = generate_keypair(p, q)
print("Public key:", public_key)
print("Private key:", private_key)

message = "HELLO"
print("Original message:", message)

encrypted_message = encrypt(message, public_key)
print("Encrypted message:", encrypted_message)

decrypted_message = decrypt(encrypted_message, private_key)
print("Decrypted message:", decrypted_message)